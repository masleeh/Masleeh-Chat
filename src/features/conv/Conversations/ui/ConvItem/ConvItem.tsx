import { Box, ListItemAvatar, ListItemButton, ListItemText, Stack, Typography } from '@mui/material';
import { AvatarEl } from 'shared/elements/common/Avatar';
import DoneIcon from '@mui/icons-material/Done';
import DoneAllIcon from '@mui/icons-material/DoneAll';
import useConvClick from '../../lib/useConvClick/useConvClick';
import { TConvItem } from 'entities/Conversations';
import useMapConvData from '../../lib/useMapConvData/useMapConvData';
import { memo } from 'react';

interface IConvItemProps {
    partData: TConvItem;
}

const ConvItem = memo(({
    partData,
}:IConvItemProps) => {
    const {handleConvClick} = useConvClick()
    const {
        conv_id,
        title,
        last_message,
        pic,
        isSelected
    } = useMapConvData(partData)

    return (
        <ListItemButton
            selected={isSelected}
            sx={{
                '&.Mui-selected': {
                    bgcolor: 'primary.main',
                    '&:hover': {
                        bgcolor: 'primary.light'
                    }
                },
                borderRadius: 4
            }}
            onClick={isSelected ? undefined : () => handleConvClick(conv_id)}
        >
            <ListItemAvatar>
                <AvatarEl src={pic} name={title ?? ''} sx={{
                    
                }}/>
            </ListItemAvatar>
            <ListItemText
                primary={
                    <Typography 
                        variant='subtitle1'
                        color={isSelected ? 'white' : 'black'}
                    >{title}</Typography>
                }
                secondaryTypographyProps={{
                    component: Box,
                }}
                secondary={
                    <Stack 
                        direction="row" 
                        justifyContent="space-between"
                        alignItems="center"
                    >
                        <Typography 
                            variant="body2"
                            color={isSelected ? 'white' : 'gray'}
                            component="span"
                        >
                            {last_message ?? ''}
                        </Typography>
                        {(title) ? (
                            <DoneAllIcon sx={{
                                color: isSelected ? 'primary.contrastText' : 'primary.light'
                            }} />
                        ) : (
                            <DoneIcon />
                        )}
                    </Stack>
                }
            />
        </ListItemButton>
    )
})

export default ConvItem