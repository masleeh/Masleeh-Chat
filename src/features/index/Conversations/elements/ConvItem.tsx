import { ListItem, ListItemAvatar, ListItemText, Stack, Typography } from '@mui/material';
import React from 'react';
import { memo } from 'react';
import { AvatarEl } from 'shared/elements/common/Avatar';
import DoneIcon from '@mui/icons-material/Done';
import DoneAllIcon from '@mui/icons-material/DoneAll';

interface IConvItemProps {
    username: string;
    last_message?: string;
    profile_pic: string | undefined;
    conv_id: string;
    isSelected: boolean;
    viewed: boolean;
}

const ConvItem = memo(({
    username,
    last_message,
    profile_pic,
    conv_id,
    isSelected,
    viewed
}:IConvItemProps) => {

    return (
        <ListItem
            sx={{
                bgcolor: (isSelected) ? 'primary.light' : 'primary.inverted'
            }}
        >
            <ListItemAvatar>
                <AvatarEl src={profile_pic} name={username} sx={{
                    
                }}/>
            </ListItemAvatar>
            <ListItemText
                primary={
                    <Typography 
                        variant='subtitle1'
                        color={isSelected ? 'white' : 'black'}
                    >{username}</Typography>}
                secondary={
                    <React.Fragment>
                        <Typography 
                            variant="body2"
                            color={isSelected ? 'white' : 'gray'}
                            component="span"
                        >
                            {last_message ?? ''}
                        </Typography>
                        {viewed ? (
                            <DoneAllIcon sx={{
                                color: isSelected ? 'primary.contrastText' : 'primary.light'
                            }} />
                        ) : (
                            <DoneIcon />
                        )}
                    </React.Fragment>
                }
            />
        </ListItem>
    )
})

export default ConvItem