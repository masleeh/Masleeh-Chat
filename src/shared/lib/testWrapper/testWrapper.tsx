import { render } from "@testing-library/react"
import { IStateSchema, TestStoreProvider } from "app/providers/store";
import i18nTest from 'shared/config/i18n/i18nTest'
import { I18nextProvider } from "react-i18next"
import { MemoryRouter } from "react-router-dom";
import { DeepPartial, EnhancedStore } from "@reduxjs/toolkit";
import { AppThemeProvider } from "app/providers/theme";

export interface componentRenderOptions {
    route?: string;
    initialState?: DeepPartial<IStateSchema>;
    store?: EnhancedStore
}

const testWrapper = (component: React.ReactNode, options: componentRenderOptions = {}) => {
    const {
        route = '/',
        initialState,
        store
    } = options

    return render(
        <TestStoreProvider initialState={initialState} store={store}>
            <AppThemeProvider>
                <MemoryRouter initialEntries={[route]}>
                    <I18nextProvider i18n={i18nTest}>
                        {component}
                    </I18nextProvider>
                </MemoryRouter>
            </AppThemeProvider>
        </TestStoreProvider>
    )
}

export default testWrapper