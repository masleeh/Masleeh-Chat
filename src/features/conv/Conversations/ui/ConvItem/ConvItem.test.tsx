import testWrapper from "shared/lib/testWrapper/testWrapper"
import ConvItem from './ConvItem'

const partData = {
    conv_id: '123456123123',
    title: undefined,
    last_message: 'ahahahaha',
    type: 'private',
    updatedAt: '2023-10-16T15:41:23.000Z',
    users: [
        {
            username: 'masleeh',
            user_id: '123',
            profile_pic: undefined
        },
        {
            username: 'katya',
            user_id: '1234',
            profile_pic: undefined
        },
    ]
}

describe('ConvItem component', () => {
    it('Check ConvItem', () => {
        const component = testWrapper(<ConvItem partData={partData}/>)
        expect(component).toMatchSnapshot()
    })
})