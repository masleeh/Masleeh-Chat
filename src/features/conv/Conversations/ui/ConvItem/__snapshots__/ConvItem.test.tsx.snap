// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConvItem component > Check ConvItem 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-10a9ai8-MuiButtonBase-root-MuiListItemButton-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-53rrgp-MuiAvatar-root"
          >
            m
          </div>
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 css-1wp00cw-MuiTypography-root"
          >
            masleeh
          </h6>
          <div
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary MuiBox-root css-pinrqx-MuiTypography-root"
          >
            <div
              class="MuiStack-root css-3y1137-MuiStack-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body2 css-98hkvb-MuiTypography-root"
              >
                ahahahaha
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-z81eeg-MuiSvgIcon-root"
                data-testid="DoneAllIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m18 7-1.41-1.41-6.34 6.34 1.41 1.41L18 7zm4.24-1.41L11.66 16.17 7.48 12l-1.41 1.41L11.66 19l12-12-1.42-1.41zM.41 13.41 6 19l1.41-1.41L1.83 12 .41 13.41z"
                />
              </svg>
            </div>
          </div>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-10a9ai8-MuiButtonBase-root-MuiListItemButton-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemAvatar-root css-1e9lk82-MuiListItemAvatar-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-53rrgp-MuiAvatar-root"
        >
          m
        </div>
      </div>
      <div
        class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-1wp00cw-MuiTypography-root"
        >
          masleeh
        </h6>
        <div
          class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary MuiBox-root css-pinrqx-MuiTypography-root"
        >
          <div
            class="MuiStack-root css-3y1137-MuiStack-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 css-98hkvb-MuiTypography-root"
            >
              ahahahaha
            </span>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-z81eeg-MuiSvgIcon-root"
              data-testid="DoneAllIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m18 7-1.41-1.41-6.34 6.34 1.41 1.41L18 7zm4.24-1.41L11.66 16.17 7.48 12l-1.41 1.41L11.66 19l12-12-1.42-1.41zM.41 13.41 6 19l1.41-1.41L1.83 12 .41 13.41z"
              />
            </svg>
          </div>
        </div>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
